### PR TITLE
@types/chrome - ScriptInjection.func args should be optional

### PR DESCRIPTION
**Pull request by davidbielik**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by davidbielik** [comment]
Sorry, I'm having trouble fixing this, but the the `func` parameter of `ScriptInjection` is showing a type error when not passing any arguments:

> Property 'args' is missing in type '{ target: { tabId: number; }; func: () => void; }' but required in type '{ args: unknown[]; }'.

```typescript
await chrome.scripting.executeScript({
    target: { tabId: 1 },
    func: function(): void {
         // do something
    }
});
```

**Comment by davidbielik** [comment]
Thanks @maxgalbu!

**Comment by maxgalbu** [comment]
In any case, `args` should be optional, correct?

@matthewkimber @otiai10 @RReverser @sreimer15 @matcarlson @ekinsol @tregagnon @EchoAbstract @spasma @bdbai @pokutuna @JasonXian @userTim @idan315 @nicolas377 @idosal 

Eg. `ScriptInjection` type should change to: 
```
export type ScriptInjection<Args extends any[] = []> = {
    /* Details specifying the target into which to inject the script. */
    target: any;
    /* The JavaScript world for a script to execute within. */
    world?: any;
} & ({
    /* The path of the JS files to inject, relative to the extension's root directory. NOTE: Currently a maximum of one file is supported. Exactly one of files and function must be specified. */
    files: string[];
} | ({
    /* A JavaScript function to inject. This function will be serialized, and then deserialized for injection. This means that any bound parameters and execution context will be lost. Exactly one of files and function must be specified. */
    func: ((...args: Args) => void);
} & {
    /* The arguments to carry into a provided function. This is only valid if the func parameter is specified. These arguments must be JSON-serializable. */
    args?: Args;
}))
```

**Comment by davidbielik** [comment]
@maxgalbu thanks for your help. That corrected `ScriptInjection` type solves the problem for me locally. lmk if i should replace my PR with that change.

File: types/chrome/index.d.ts [hunk 1]
```diff
@@ -7593,22 +7593,19 @@ declare namespace chrome.scripting {
 
     export type ScriptInjection<Args extends any[] = []> = {
         /* Details specifying the target into which to inject the script. */
-        target: InjectionTarget;
+        target: any;
```
Comment (davidbielik) [inline types/chrome/index.d.ts:7596]:
```suggestion
        target: InjectionTarget;
```

File: types/chrome/index.d.ts [hunk 2]
```diff
@@ -7593,22 +7593,19 @@ declare namespace chrome.scripting {
...
         /* The JavaScript world for a script to execute within. */
-        world?: ExecutionWorld;
+        world?: any;
```
Comment (davidbielik) [inline types/chrome/index.d.ts:7598]:
```suggestion
        world?: InjectionTarget;
```

File: types/chrome/index.d.ts [hunk 3]
```diff
@@ -7595,20 +7595,17 @@ declare namespace chrome.scripting {
         /* Details specifying the target into which to inject the script. */
         target: InjectionTarget;
         /* The JavaScript world for a script to execute within. */
-        world?: ExecutionWorld;
+        world?: InjectionTarget;
```
Comment (davidbielik) [inline types/chrome/index.d.ts:7598]:
```suggestion
        world?: ExecutionWorld;
```

**Comment by maxgalbu** [comment]
yep sorry for the `any`, i was testing on the typescript playground haha

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by nicolas377** [comment]
Sorry I'm so late, this seems like a bad idea since a function with args should have args passed to it. This would allow for someone to define a function with args and not pass any.

**Comment by maxgalbu** [comment]
@nicolas377 Would something like this work?

```
export type ScriptInjection<Args extends any[]> = {
    /* Details specifying the target into which to inject the script. */
    target: InjectionTarget;
    /* The JavaScript world for a script to execute within. */
    world?: ExecutionWorld;
} & ({
    /* The path of the JS files to inject, relative to the extension's root directory. NOTE: Currently a maximum of one file is supported. Exactly one of files and function must be specified. */
    files: string[];
} | ({
    /* A JavaScript function to inject. This function will be serialized, and then deserialized for injection. This means that any bound parameters and execution context will be lost. Exactly one of files and function must be specified. */
    func: () => void;
}  | {
    /* A JavaScript function to inject. This function will be serialized, and then deserialized for injection. This means that any bound parameters and execution context will be lost. Exactly one of files and function must be specified. */
    func: (...args: Args) => void;
    /* The arguments to carry into a provided function. This is only valid if the func parameter is specified. These arguments must be JSON-serializable. */
    args: Args;
}))
``` 

see here for some tests: https://tsplay.dev/WzO9kw

**Comment by nicolas377** [comment]
That looks good to me! I added some more tests [here](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcDKBjKBLMMCSAdgFbAYzYSEBiUEAtgAoCuANqwErACOzwAzjAA8AQSgBzfnFAxghACZSAhoUQBtALoA+OAF44AbwBQcOAHoAVHAAiwGEuysp-FBmwAzRNkLiEAC1R7CTs4bxgIOAB3P2wMPwQI71Jyf1R+LFwYADo4CzMTBCVgmAAuOBVEAG4CyzgAFQC4ACklADclTBw8KOhWeTh3aHK4dK74cOkQMmZZKOwYGMIcvILI3vkAfjKK6oBfOAAyOAAKAoAfQwLTU1qG1DAlBbgId1TmtAHHAQTQkjIYAA0cCgwFYj2wrUCEQWqBkcn4lEIAHIpHQIPB5NgQeRoIgcgA5ADydQAomUAMLMKAgwgwViIYb0JQgbD0Zj0Z6vKiodxfULOZhgSCwYDyHIkkBKcj056EVAvT6sb4qfruZiEciIuBswRwABGaVcHmwouW+Wu115Sv4ZUEOB8mmqFt25xOV2uF2MFu95isImabQ6GW6ao1FCoPyS-xyDWwUlDmojkUcrH1aWAOCUrGwAC9RUCValCHB5AIM9gs7nRQMhlHE0t6jEpPRgCopAtHuVVPqIOr+g8oEoW7IoMoFJNpuHixgqLIQPBk+w03BWBBBOLJdKGdzOYrleOE1PtcxdQaRkbeabcuafZb1RgyscAJR6HStCDYeRO28un2e90+rU-otO0nSZAM95HhMdbZI2cYQWGWqLqmZ78OWlZ5vIBbjjCxalmhmbZphNZQL8yRTjGTbaq2hDtn4nYVD2fZwAOQ52BmY79KAk5ajOtIyHMS5nqu65wBKUp0tucq7la+6qpBWo6vAqEXiaYrXgB3oJo+WS6UUkhlGIkgvrob4fl+mkWrcjT6eycgwO2EQYEUUAMmEERKCxdCtJ+1aHoilHwfBVAyu02b9B4bwJixRRsSO-LnmQxpXncaHlBIdm0s2J7KagTRoIS+IALQERWRFKHqSpmpZ1z6TacBGfw34+r+1xPkYT7VEYoDCvA-kRtxGAzMAYF4LQDAsOwXC8AIwiNZMsgKGO6jaKcpgwYiZSjQQfz1uNTBsJwPB8IIogSPwWgdWU76foYLpGGYZhoH4vZ9AMDisN1UxDbI237ZNR0zYIxxeoUxRlAARLIggQwCBTaQh9YnNDpQjDA9riC+Bi7HDuydQ9T0vWwqofV9k4jcGMD-Yd00nTAIMFEE4h2JDKOwwUdVlJocOmAj-XFscKO2uj3iY3duP449z2vSTjhkz9FNjNTU3HbNDOmEzLNwFDs3s6YnNwGoEP8AwdiLOIEMaDziNlPzJxYzjRh411UtE29RJ1O9cuDcNf10AdKtA-ToOa6jOsw9bfMKRGz7i07+ME9LxNe59Pu-ZTyuA3T6tg8zYds9bBvc-D9629HAtC2jGMOxLLuEzLcAeyn8u+xn-sA7Tash-pWvhzAevpQZht95bkel4jR6C7NwvV3HztGEAA). Also, remember to implement tests!

**Review by nicolas377 — CHANGES_REQUESTED** [review]
Holding this PR until previously mentioned changes are made.

**Comment by nicolas377** [comment]
Actually, I'm pretty confident that with those changes, this PR is just a reimplementation and brings no changes, so it's unnecessary.

**Comment by maxgalbu** [comment]
@nicolas377 I don't really understand. Your suggested changes are different from what we talked about and tested on the typescript playground 🤔 

These changes surely bring an improvement, because as we said at the beginning of the pull request, this should not be an error, but it's an error with the current type:

```ts
//Should NOT fail
executeScript({
    target: "test",
    func: function() {

    }
})
```

**Comment by nicolas377** [comment]
Right! Sotty about that. I'm good with the type in the playground link i sent back. Just remember to add tests, and I'm good!

**Comment by maxgalbu** [comment]
I just looked up tests and it already looks like they don't fail:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/81ae50d8fa0ae4e78ecb8a8dbbe5c22f479491dc/types/chrome/test/index.ts#L983

But.... why this doesn't fail with the current type? 

And, if I add this line:
```ts
await chrome.scripting.executeScript({ target: { tabId: 0 }, func: function() {} });
```

...then it fails. wtf?

**Comment by nicolas377** [comment]
That might be an upstream issue with typescript. Could you make a playground for that?

**Comment by maxgalbu** [comment]
[Here](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcDKBjKBLMMCSAdgFbAYzYSEBiUEAtgAoCuANqwErACOzwAzjAA8AQSgBzfnFAxghACZSAhoUQBtALoA+OAF44AbwBQcU3AD0AKjgARYDCXZWU-igzYAZomyFxCABaoDhL2cD4wEHAA7v7YGP4IkT6k5AGo-Fi4MAB0cJbmJmbB4vYAXHAqiADchaZWcAAqgXAAUkoAbkqYOHjR0KzycB7QFXAZPfAR0iBkzLLR2DCxhLn5tX1QAwD85ZU1AL5wAGRwABTGZhbWTahgSktwEB5prWhDTgKJYSRkMAA0cCgwFY92w7SCkSWqBkcn4lEIAHIpHQIPB5NggeRoIhcgA5ADyDQAouUAMLMKBAwgwViIUb0JQgbD0Zj0R7PKioDwfMIuZhgSCwYDyXJEkBKci0x6EVBPd6sT4qQYeZiEcjwuAswRwABG6TcnmwwtWBUu3IV-HKghwvk0BzgAB8zhczPURK0Ol1Mr0VWqKFQvslfrkmtgpL71QGok5WLr0sAcEpWNgAF7CgFKtKEODyAQJ7BJ1PCoYjIORlaNWJSejAFRSJb3CqqXUQVWDO5QJQ12RQZQKaazf3ZjBUWQgeDR9hxuCsCCCUXiyV0zns+WK-sRoea5javVjA3c415U1mCPlU4ASj0OnaEGw8ntjsM6zdHs63SyQ1V5cDP3IIarL8-Q1SdYz3fh80LNN5AzfsoWzXMIMTZNoJLKBvhSIcALDTVa0Iet-EbSoWzbOAOy7ewEz7QZQEHDUR2pGQFinPdZ3nOAxQlGllxlVdzXXZVvy3LV4HAg8jRFY91jPM5sjkpQJEtOAxEkK9dBvO8Hxfa5mgU8RWTkGB60iDAFKgOlwkiJQyLodp72LTd4WwqQcKoKVOmTQZPBeCMyIUiie15fcyENI8bggioJAM6lqx3UTUBaNB8VxABaJCCxQpQdQVE11j0pSVP4A4LwvIwjFAQVJmQVAPzwIhMPhfEBlERTplkBQ+3UDQ9DgTQdH0F06msOwHCcFwD28XwXmKUJLOiWJ4l-TCXnGLJcsuGaYF2VQakueobjfL0Jg2AY0NGVbeimWiMDmVBoyWHx1rMKJ+nkHYm2qIxDhOc5tMrW57gSOUoVeNdjIw34ASBEEKHBL4QZhfD4SRQEIFRHMMV+bE8UJEk4HJSlDKlayGSZFk2TlFd+KC-h+Uqo9OKXaVZWefjqKAn8ROnVwQsPSS1jND4lOtHxxDtL6n1+vbrHdNp329eBHIDKYyxySscKV7NQO5yCUPTJtBngnM82QotlVLP8sPV6s8IIojmx1Vt+3I7sqINgcbq3BixwnGNpzYtXGe45m+KF93Ne3Xd9V5iSntMGTTjk7J8vKQq1I0+97R+wq2rkRReu6rZn2l-7Iv0msYq+UzKQs6krJsiA7NzQTgKoZygrcukPPvMJnhB3yXco9CcJ59w+YAvMy+iozI-i15krS3XUyynKpMufL3sK+1ykGq5S706fwer8zvimaywFs+yW-LdvXMIdzCy8vvmgH-zXeH8aY7CwIIoPiuZ65nuRKC90pQRXsAOOZcCqKWKiVMqFVoCKyEhqa6t1aowCavIFqkhc4dQ+n1U4qt4TlHQfVX4jVmqFS0Becot4e7GC+uVcAiCOZblQbIdBtAGAsHYFwXgAhhA50RvnSoBCiFUBIQrMh5YuFMDYJwHgfBBDYP4NQ2hmlnyMLKkYcw5hSGWwoYbAR-Ayq6LQP4Vsp0PCOFYEwwcwB0GYKlkUPSZQ4AACJZCCHcX8PKilyiaF8Wab85RNanC8VtEWvgrwMMuPsL6pUzEWLYMqGxdjPYOIVk43em1yieIET46SITWHwnCQI0oUTxAxLKnEhJOi9HJKsWk9hmSJjZPWLkjxETCnr38b1dx-AGD2GWOIdxGggmnmKWE6p6x4n7ESQ0yxgwCQNCGM0mYGTHEDGcaYTp+TvETPjlM5BVBLyaNqfM+p5illwBWWspw6S0FZO2Tk1xW0ukFMORzc86dzlmDmaVUxeipEGKoLInhCj+HagiSYq5jTUkPJaZwugcjeGKIETshAby8ndK+TJMJESKkwBtFUv5pgAVwpudYxFGynkTHBfIvhSiYCYr2bivxkgAnjKKWqUJJzCBlMEESklMyLkLOuSk+5tikUKwZWiqFLLXkhHefsmAPSzAp36YM7sIyxl4uOa3AVor-l1KSTcu51LpW0o4bKlFEKmUYqVSUFV7Lgm8pKac415K6kyvpXaxl6LBCsuxR8g5PKMA-OvGSuAAKgA)'s an added test. Using function() is an error with the old type, but not with the pull request type.

**Comment by davidbielik** [comment]
Hi @maxgalbu thanks for your help. This is a little over my head, so I'll close this and wait for your separate PR to fix this issue.

**Pull request closed without merging**